### DISCUSSION TITLE
Unexpected alignment error

### DISCUSSION THREAD

**Discussion by euro-phd** [EduMIPS64/edumips64#701] [Q&A] [answered]
Hi, I'm studying MIPS64 at uni as part of the Computer Architecture course. One of the labs requires us to execute the following code:

```
.text
div.d f7,f9,f10
mul.d f2,f4,f3
sub.d f7,f7,f4
ld r1,78(r0)
add.d f4,f5,f6
halt
```

During the execution, using EduMIPS64, the error occurs.
![edumips64](https://user-images.githubusercontent.com/92214233/187075131-8d657c31-06cf-494c-b40b-3cc42ff3e879.png)

However, on WinMIPS64, used by a lecturer, everything runs smoothly.
![winmips64](https://user-images.githubusercontent.com/92214233/187075136-8de0d7ea-9479-447a-96be-1e67d4ae355a.png)

Is there a way to patch/reconfigure EduMIPS64 to receive the same results as WinMIPS64? It seems like WinMIPS64 [fixed alignment on its own](https://github.com/AndoniZubimendi/WinMIPS64/blob/d1fd07fb4655d2d7d6b47f38f038210a50af301a/src/utils.cpp). However, I'm unsure how to code the same behavior in EduMIPS64.

Thanks

**Comment by lupino3** [comment, marked as answer]
Hi @euro-phd, thanks a lot for raising this issue!

I think that in this case EduMIPS64 deals with misaligned load instructions correctly.
If you look at the the MIPS64 ISA documentation (https://s3-eu-west-1.amazonaws.com/downloads-mips/documents/MIPS_Architecture_MIPS64_InstructionSet_%20AFP_P_MD00087_06.05.pdf), you'll see that `LD` requires the operand to be aligned to 8 bytes. (we implemented this before release 6)

<img width="488" alt="image" src="https://user-images.githubusercontent.com/200139/187077035-e60a5c39-2c2a-4562-999d-f13d4b988924.png">

When we implemented load/store, we also saw that there were patents for handling non-aligned memory accesses, so we decided to only support aligned memory accesses. See https://github.com/EduMIPS64/edumips64/issues/2.

In your example, given that the memory is empty, is there a chance you can run

```
ld r1, 72(r0)
```

or 

```
ld r1, 80(r0)
```

instead of the instruction that causes a problem?

Hope this helps, thanks!

**Reply by lupino3** [reply]
This is also documented in the manual: https://edumips64.readthedocs.io/en/latest/instructions.html#load-store-instructions
<img width="926" alt="image" src="https://user-images.githubusercontent.com/200139/187077225-e6e80dc6-cd76-4cc3-ac91-20b0e775bd97.png">

**Comment by lupino3** [comment]
HI, I'm not sure exactly how much work would be required. Every load/store
instruction needs to be modified.

The semantics are in Appendix B of the most recent MIPS64 ISA spec, I have
created an issue to track your request and I'll try to investigate the
implementation it as time permits:
https://github.com/EduMIPS64/edumips64/issues/703

Thanks,
Andrea

Il giorno dom 28 ago 2022 alle ore 16:20 wa1tf0r.me <
***@***.***> ha scritto:

> Thank you for answering my question. Now I understand more about MIPS64
> and how to work with it.
> Nevertheless, I wonder how many changes are needed to handle non-aligned
> memory accesses, just as an option that can be turned on in settings.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/EduMIPS64/edumips64/discussions/701#discussioncomment-3493266>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABQ3S2BTN3M7GFBGSGZAGLV3NYSVANCNFSM573BQ2EA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Andrea Spadaccini, Ph.D.
About me: https://about.me/andreaspadaccini
a.k.a. lupino3 / spadaccio
